src/e2e_tests.rs

```diff
@@ -0,0 +1,649 @@
+        address: "127.0.0.1:5860".to_string(),
+        entities: vec![],
+    };
+    let twin3: Twin = Twin {
+        version: 1,
+        id: 3,
+        account: t3.account(),
+        address: "127.0.0.1:5870".to_string(),
+        entities: vec![],
+    };
+
+    // creating mock db with dummy entities
+    let db1 = create_mock_db(vec![&twin1, &twin2, &twin3]).unwrap();
+    let db2 = create_mock_db(vec![&twin1, &twin2, &twin3]).unwrap();
+    let db3 = create_mock_db(vec![&twin1, &twin2, &twin3]).unwrap();
+
+    // start redis servers
+    let redis_manager = RedisManager {
+        ports: vec![local_redis_port, remote1_redis_port, remote2_redis_port],
+    };
+    redis_manager.init().await;
+    // create redis storage
+    let storage1 = create_local_redis_storage(local_redis_port).await.unwrap();
+    let storage2 = create_local_redis_storage(remote1_redis_port)
+        .await
+        .unwrap();
+    let storage3 = create_local_redis_storage(remote2_redis_port)
+        .await
+        .unwrap();
+
+    // start three instances of RMB
+    tokio::spawn(async move { start_rmb(db1, storage1, t1, &twin1.address, twin1.id).await });
+    tokio::spawn(async move { start_rmb(db2, storage2, t2, &twin2.address, twin2.id).await });
+    tokio::spawn(async move { start_rmb(db3, storage3, t3, &twin3.address, twin3.id).await });
+    // mimic a process that handle a command `testme` from two remote nodes
+    let cmd = "testme";
+    tokio::spawn(async move { handle_cmd(cmd, remote1_redis_port).await.unwrap() });
+    tokio::spawn(async move { handle_cmd(cmd, remote2_redis_port).await.unwrap() });
+    // test simple message exchange
+    let (_responses, err_count, _success_count, timed_out_count) =
+        send_and_wait(cmd, vec![twin2.id, twin3.id], 1, local_redis_port).await;
+    assert_eq!(err_count, 0);
+    assert_eq!(timed_out_count, 0);
+    let (_responses, err_count, _success_count, timed_out_count) =
+        send_and_wait(cmd, vec![twin2.id, twin3.id], 200, local_redis_port).await;
+    assert_eq!(err_count, 0);
+    assert_eq!(timed_out_count, 0);
+    let (_responses, err_count, _success_count, timed_out_count) =
+        send_and_wait(cmd, vec![twin2.id, twin3.id], 1000, local_redis_port).await;
+    assert_eq!(err_count, 0);
+    assert_eq!(timed_out_count, 0);
+}
+
+#[tokio::test]
+async fn test_twin_not_found() {
+    initialize_logger();
+    let local_redis_port = 6387;
+    let t1 = create_test_signer(KPair::SrPair).unwrap();
+
+    // create dummy entities for testing
+    let twin1: Twin = Twin {
+        version: 1,
+        id: 1,
+        account: t1.account(),
+        address: "127.0.0.1:5880".to_string(),
+        entities: vec![],
+    };
+
+    // creating mock db with dummy entities
+    let db1 = create_mock_db(vec![]).unwrap();
+
+    // start redis servers
+    let redis_manager = RedisManager {
+        ports: vec![local_redis_port],
+    };
+    redis_manager.init().await;
```

Comment:
In general in `rust` literal struct creation is not used like that. instead this can be something like
```
let redis_manager = RedisManager::new(ports).await;
```

The new then is the one that creates the Manager and `initialize` it.
Usually the struct filds are going to be `private` hence you can only construct it via public methods.

Reply:
Thanks, I'll implement an associated function to create the struct, but will keep `init` as a separate method to start the Redis instances, to ensure a proper dropping/teardown if a problem happened in the middle of starting the instances. this wouldn't possible or straightforward if we proceed with starting the instances before creating the dropable struct. please resolve if you agree.